File: moonpoison/stack/735.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& ast) 
    {
        stack<int> pos, neg;
        vector<int> ans;
        for(int i=0;i<ast.size();++i)
        {
            if(ast[i]>0) pos.push(ast[i]); //if pos
            else //if neg
            {
                if(pos.empty()) ans.push_back(ast[i]); //if pos stack is empty
                else //if pos stack isnt empty
                {
                    neg.push(ast[i]);
                    while(!pos.empty()&&!neg.empty())
                    {
                        if(abs(pos.top())>abs(neg.top())) neg.pop();
                        else if(abs(pos.top())==abs(neg.top()))
                        {
                            pos.pop();
                            neg.pop();
                        }
                        else pos.pop();
                    }
                    if(!neg.empty())
                    {
                        ans.push_back(neg.top());
                        neg.pop();
                    }
                }
            }
        }
        if(!pos.empty())
        {
            vector<int> v;
            while(!pos.empty())
            {
                v.push_back(pos.top());
                pos.pop();
            }
            reverse(v.begin(), v.end());
            ans.insert(ans.end(), v.begin(), v.end());
        }
        return ans;
    }
};

File: ice0208/stack/735.py
class Solution:
    def asteroidCollision(self, asteroids: list[int]) -> list[int]:
        stack = []

        for a in asteroids:
            if a > 0:
                stack.append(a)
            else:
                self.collide(stack, a)

        return stack

    def collide(self, stack, a):
        while True:
            if len(stack) == 0:
                stack.append(a)
                return
            elif stack[-1] < 0:
                stack.append(a)
                return
            else:
                if stack[-1] == -a:
                    stack.pop()
                    return
                elif stack[-1] < -a:
                    stack.pop()
                else:
                    return


if __name__ == "__main__":

    testcase_list = [[5, 10, -5], [8, -8], [10, 2, -5]]

    for testcase in testcase_list:
        print(Solution().asteroidCollision(asteroids=testcase))
